File: 1001-2000/1339. Maximum Product of Splitted Binary Tree.java
// Medium
// DFS
// Time:O(n),Space:O(h)
// https://leetcode.cn/problems/maximum-product-of-splitted-binary-tree/

class Solution {
  // DFS to calculate the totalSum of tree, and calculate the subtreeSum
  //  so the remain part is totalSum-subtreeSum, then calculate and update  
  //  the max product= subtreeSum * (totalSum - subtreeSum)
  private long maxSum = 0;
  private long totalSum = 0;

  public int maxProduct(TreeNode root) {
      totalSum = getTotal(root); 
      getTotal(root); // get subtreeSum
      return (int)(maxSum % (1_000_000_007)); // get mod
  }
  private long getTotal(TreeNode root) {
      if (root == null) return 0;
      long leftSum = getTotal(root.left);
      long rightSum = getTotal(root.right);
      long subtreeSum = root.val + leftSum + rightSum;

      // calculate and update the maxSum
      if (totalSum != 0) {
          maxSum = Math.max(maxSum, subtreeSum * (totalSum - subtreeSum));
      }
      return subtreeSum;
  }
}

File: 1001-2000/1339. Maximum Product of Splitted Binary Tree.py
# Time:O(n), Space:O(h)

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        MOD = 10**9 + 7

        # calculate total
        def get_sum(node):
            if not node:
                return 0
            return node.val + get_sum(node.left) + get_sum(node.right)

        total = get_sum(root)
        max_prod = 0

        def dfs(node):
            nonlocal max_prod
            if not node:
                return 0

            sub_sum = node.val + dfs(node.left) + dfs(node.right)
            prod = sub_sum * (total - sub_sum)
            max_prod = max(max_prod, prod)

            return sub_sum

        dfs(root)
        return max_prod % MOD
    